relay/polkadot/src/lib.rs

```diff
@@ -1689,27 +1698,111 @@ pub type Migrations = migrations::Unreleased;
+				return Ok(Vec::new())
+			}
+
+			log::info!(target: "runtime::session_keys", "Collecting pre-upgrade session keys state");
+			let key_ids: Vec<_> = SessionKeys::key_ids()
+				.into_iter()
+				.filter(|&k| *k != sp_core::crypto::key_types::BEEFY)
+				.collect();
+			frame_support::ensure!(
+				key_ids
+					.clone()
+					.into_iter()
+					.find(|&k| *k == sp_core::crypto::key_types::IM_ONLINE) ==
+					None,
+				"New session keys contain the ImOnline key that should have been removed",
+			);
+			let storage_key = pallet_session::QueuedKeys::<Runtime>::hashed_key();
+			let mut state: Vec<u8> = Vec::new();
+			frame_support::storage::unhashed::get::<Vec<(ValidatorId, OldSessionKeys)>>(
+				&storage_key,
+			)
+			.ok_or::<sp_runtime::TryRuntimeError>("Queued keys are not available".into())?
+			.into_iter()
+			.for_each(|(id, keys)| {
+				state.extend_from_slice(id.as_slice());
+				for key_id in key_ids.clone() {
+					state.extend_from_slice(keys.get_raw(*key_id));
+				}
+			});
+			frame_support::ensure!(state.len() > 0, "Queued keys are not empty before upgrade");
+			Ok(state)
+		}
+
 		fn on_runtime_upgrade() -> Weight {
+			if System::last_runtime_upgrade_spec_version() > UPGRADE_SESSION_KEYS_FROM_SPEC {
+				log::info!("Skipping session keys upgrade: already applied");
+				return <Runtime as frame_system::Config>::DbWeight::get().reads(1)
+			}
+			log::trace!("Upgrading session keys");
 			Session::upgrade_keys::<OldSessionKeys, _>(transform_session_keys);
 			Perbill::from_percent(50) * BlockWeights::get().max_block
 		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(
+			old_state: sp_std::vec::Vec<u8>,
+		) -> Result<(), sp_runtime::TryRuntimeError> {
+			if System::last_runtime_upgrade_spec_version() > UPGRADE_SESSION_KEYS_FROM_SPEC {
+				log::warn!(target: "runtime::session_keys", "Skipping session keys migration post-upgrade check due to spec version (already applied?)");
+				return Ok(())
+			}
+
+			let key_ids: Vec<_> = SessionKeys::key_ids()
+				.into_iter()
+				.filter(|&k| *k != sp_core::crypto::key_types::BEEFY)
+				.collect();
+			let mut new_state: Vec<u8> = Vec::new();
+			pallet_session::QueuedKeys::<Runtime>::get().into_iter().for_each(|(id, keys)| {
+				new_state.extend_from_slice(id.as_slice());
+				for key_id in key_ids.clone() {
+					new_state.extend_from_slice(keys.get_raw(*key_id));
+				}
+			});
+			frame_support::ensure!(new_state.len() > 0, "Queued keys are not empty after upgrade");
+			frame_support::ensure!(
+				old_state == new_state,
+				"Pre-upgrade and post-upgrade keys do not match!"
+			);
+			log::info!(target: "runtime::session_keys", "Session keys migrated successfully");
+			Ok(())
+		}
 	}
 
 	/// Unreleased migrations. Add new ones here:
 	pub type Unreleased = (
-		// Upgrade SessionKeys to include BEEFY key
+		// Upgrade SessionKeys to exclude ImOnline key and include BEEFY
 		UpgradeSessionKeys,
```

Comment:
beefy upgrade has been already applied in 1.1.0: https://github.com/polkadot-fellows/runtimes/releases/tag/v1.1.0
<img width="844" alt="beefy-polkadot" src="https://github.com/polkadot-fellows/runtimes/assets/4211399/4998bf3c-f906-47b9-882e-3da57d6ef967">
we need to remove it from unreleased

Reply:
i hope we didn't add it to subsequent releases again 🙈

Reply:
looks like we skipped some releases and applied 1.1.2 directly

Reply:
> looks like we skipped some releases and applied 1.1.2 directly
> 
> but given how easy it is to overlook this `UpgradeSessionKeys` slipping into the next release, we should be more careful, not forget removing this in the next release and do something about it [paritytech/polkadot-sdk#2421](https://github.com/paritytech/polkadot-sdk/issues/2421)

We won't forget, it is one of todo points for the next release: https://github.com/polkadot-fellows/runtimes/issues/140

Reply:
And even if we forget, the migration is guarded by the spec version: https://github.com/polkadot-fellows/runtimes/pull/178/files#diff-efa4caeb17487ecb13d8f5eb7863c3241d84afa2e73fbf25909a2ca89df0f362R1751